rsqueakvm/interpreter.py

```diff
@@ -304,27 +318,60 @@ def getreceiverclass(self, s_context):
         return s_context.w_receiver().safe_getclass(self.space)
 
     def getblockmethod(self, s_context):
-        return s_context.blockmethod
+        return s_context.blockmethod()
 
     def loop_bytecodes(self, s_context, may_context_switch=True):
         old_pc = 0
+        jump_back_pc = 0
+        unrollings = 0
         if not jit.we_are_jitted() and may_context_switch:
             self.quick_check_for_interrupt(s_context)
         method = s_context.w_method()
+        frame_size = method.frame_size()
         while True:
             pc = s_context.pc()
             if pc < old_pc:
-                if jit.we_are_jitted():
-                    # Do the interrupt-check at the end of a loop, don't interrupt loops midway.
-                    self.jitted_check_for_interrupt(s_context)
-                self.jit_driver.can_enter_jit(
-                    pc=pc, self=self, method=method,
-                    w_class=self.getreceiverclass(s_context),
-                    blockmethod=self.getblockmethod(s_context),
-                    s_context=s_context)
+                if jump_back_pc == old_pc:
+                    if (unrollings < self.optargs.max_squeak_unroll_count and
+                        jit.current_trace_length() < self.optargs.squeak_unroll_trace_limit):
+                        unrollings += 1
+                    else:
+                        if jit.we_are_jitted():
+                            # Do the interrupt-check at the end of a loop, don't
+                            # interrupt loops midway.
+                            self.jitted_check_for_interrupt(s_context)
+                        if not s_context.has_overflow_stack():
```

Comment:
When we're in a loop where the initially allocated frame size was smaller than what we actually used in the loop, we'll have an overflow stack. The next time we enter this method the larger frame size will be allocated and no overflow stack will be needed, but we don't want to jit a loop that uses the overflow stack only for the one time we have it.